File: java/lib/src/main/java/io/v/v23/security/BlessingRoots.java
// Copyright 2015 The Vanadium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package io.v.v23.security;

import com.google.common.collect.Multimap;
import io.v.v23.verror.VException;

import java.security.interfaces.ECPublicKey;

/**
 * The set of authoritative public keys for roots of blessings.
 * <p>
 * See also: <a href="https://vanadium.github.io/glossary.html#blessing-root">https://vanadium.github.io/glossary.html#blessing-root</a>.
 */
public interface BlessingRoots {
    /**
     * Marks {@code root} as an authoritative key for blessings that match {@code pattern}.
     *
     * Multiple keys can be added for the same pattern, in which case all those keys are considered
     * authoritative for blessings that match the pattern.
     *
     * @param  root            root that is deemed an authoritiative key for the provided pattern
     * @param  pattern         pattern for which we're assigning the authoratitative key
     * @throws VException      if there was an error assigning the root
     */
    void add(ECPublicKey root, BlessingPattern pattern) throws VException;

    /**
     * Returns {@code null} iff the provided root is recognized as an authority on a pattern
     * that is matched by the blessing.
     *
     * @param  root            the root key we're checking for authority
     * @param  blessing        the blessing we're checking against the root
     * @throws VException      if the provided root is not recognized as an authority for the
     *                         provided blessing
     */
    void recognized(ECPublicKey root, String blessing) throws VException;

    /**
     * Returns the set of recognized roots as a map from {@link BlessingPattern} to the set
     * of authoritative keys for that pattern.
     *
     * @throws VException      if there was an error preparing the dump
     */
    Multimap<BlessingPattern, ECPublicKey> dump() throws VException;

    /**
     * Return a human-readable string description of the roots.  This description is detailed and
     * lists out the contents of the roots.  Use {@link Object#toString} method for a more
     * succinct description.
     *
     * @return human-readable string description of the roots
     */
    String debugString();
}

File: go/src/v.io/x/ref/lib/security/blessingroots.go
// Copyright 2015 The Vanadium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package security

import (
	"bytes"
	"fmt"
	"sort"
	"sync"

	"v.io/v23/security"
	"v.io/v23/verror"
	"v.io/x/lib/vlog"
	"v.io/x/ref/lib/security/serialization"
)

var errRootsAddPattern = verror.Register(pkgPath+".errRootsAddPattern", verror.NoRetry, "{1:}{2:} a root cannot be recognized for all blessing names (i.e., the pattern '...')")

// blessingRoots implements security.BlessingRoots.
type blessingRoots struct {
	persistedData SerializerReaderWriter
	signer        serialization.Signer
	mu            sync.RWMutex
	state         blessingRootsState // GUARDED_BY(mu)
}

func (br *blessingRoots) Add(root []byte, pattern security.BlessingPattern) error {
	if pattern == security.AllPrincipals {
		return verror.New(errRootsAddPattern, nil)
	}
	// Sanity check to avoid invalid keys being added.
	if _, err := security.UnmarshalPublicKey(root); err != nil {
		return err
	}
	key := string(root)
	br.mu.Lock()
	defer br.mu.Unlock()
	patterns := br.state[key]
	for _, p := range patterns {
		if p == pattern {
			return nil
		}
	}
	br.state[key] = append(patterns, pattern)

	if err := br.save(); err != nil {
		br.state[key] = patterns[:len(patterns)-1]
		return err
	}

	return nil
}

func (br *blessingRoots) Recognized(root []byte, blessing string) error {
	key := string(root)
	br.mu.RLock()
	for _, p := range br.state[key] {
		if p.MatchedBy(blessing) {
			br.mu.RUnlock()
			return nil
		}
	}
	br.mu.RUnlock()
	// Silly to have to unmarshal the public key on an error.
	// Change the error message to not require that?
	obj, err := security.UnmarshalPublicKey(root)
	if err != nil {
		return err
	}
	return security.NewErrUnrecognizedRoot(nil, obj.String(), nil)
}

func (br *blessingRoots) Dump() map[security.BlessingPattern][]security.PublicKey {
	dump := make(map[security.BlessingPattern][]security.PublicKey)
	br.mu.RLock()
	defer br.mu.RUnlock()
	for keyStr, patterns := range br.state {
		key, err := security.UnmarshalPublicKey([]byte(keyStr))
		if err != nil {
			vlog.Errorf("security.UnmarshalPublicKey(%v) returned error: %v", []byte(keyStr), err)
			return nil
		}
		for _, p := range patterns {
			dump[p] = append(dump[p], key)
		}
	}
	return dump
}

// DebugString return a human-readable string encoding of the roots
// DebugString encodes all roots into a string in the following
// format
//
// Public key     Pattern
// <public key>   <patterns>
// ...
// <public key>   <patterns>
func (br *blessingRoots) DebugString() string {
	const format = "%-47s   %s\n"
	b := bytes.NewBufferString(fmt.Sprintf(format, "Public key", "Pattern"))
	var s rootSorter
	for keyBytes, patterns := range br.state {
		key, err := security.UnmarshalPublicKey([]byte(keyBytes))
		if err != nil {
			return fmt.Sprintf("failed to decode public key: %v", err)
		}
		s = append(s, &root{key, fmt.Sprintf("%v", patterns)})
	}
	sort.Sort(s)
	for _, r := range s {
		b.WriteString(fmt.Sprintf(format, r.key, r.patterns))
	}
	return b.String()
}

type root struct {
	key      security.PublicKey
	patterns string
}

type rootSorter []*root

func (s rootSorter) Len() int           { return len(s) }
func (s rootSorter) Less(i, j int) bool { return s[i].patterns < s[j].patterns }
func (s rootSorter) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }

func (br *blessingRoots) save() error {
	if (br.signer == nil) && (br.persistedData == nil) {
		return nil
	}
	data, signature, err := br.persistedData.Writers()
	if err != nil {
		return err
	}
	return encodeAndStore(br.state, data, signature, br.signer)
}

// NewBlessingRoots returns an implementation of security.BlessingRoots
// that keeps all state in memory.
func NewBlessingRoots() security.BlessingRoots {
	return newInMemoryBlessingRoots()
}

// newInMemoryBlessingRoots returns an in-memory security.BlessingRoots.
//
// The returned BlessingRoots is initialized with an empty set of keys.
func newInMemoryBlessingRoots() security.BlessingRoots {
	return &blessingRoots{
		state: make(blessingRootsState),
	}
}

// newPersistingBlessingRoots returns a security.BlessingRoots for a principal
// that is initialized with the persisted data. The returned security.BlessingRoots
// also persists any updates to its state.
func newPersistingBlessingRoots(persistedData SerializerReaderWriter, signer serialization.Signer) (security.BlessingRoots, error) {
	if persistedData == nil || signer == nil {
		return nil, verror.New(errDataOrSignerUnspecified, nil)
	}
	br := &blessingRoots{
		state:         make(blessingRootsState),
		persistedData: persistedData,
		signer:        signer,
	}
	data, signature, err := br.persistedData.Readers()
	if err != nil {
		return nil, err
	}
	if (data != nil) && (signature != nil) {
		if err := decodeFromStorage(&br.state, data, signature, br.signer.PublicKey()); err != nil {
			return nil, err
		}
	}
	return br, nil
}
